src/firewall/functions.py

```diff
@@ -26,6 +33,41 @@
 }
 
 
+def str_to_bool(val, default=DEFAULT_SENTINEL, ignore_error=False):
```

Comment:
The API is the way it is, because it makes sense. We want to have one function for parsing a boolean value. And that needs to give us the flexibility to accept/reject bogus values, and to determine the default value.

Note that currently `str_to_bool()` has yet almost no users. So of course, not all features are currently in use. When you `git grep '"yes"' origin/main -- 'src/*.py*'`, you an see that there are many places that re-implement str-to-bool (badly and inconsistently). `str_to_bool()` is supposed to become *the* true way.

So:

1) if you just call `str_to_bool(val)`, then the function will raise an exception unless it can clearly interpret `val` as truthy/falsy. Otherwise, the function returns `True` or `False`. That is the strict parsing mode, that you get by default.

2) if you do `str_to_bool(val, default)`, then `None`/`""`/`"default"`/`"-1"` are additionally accepted. In that case, `default` will be returned. Which the user can set to anything, including `default=None`. This is still strict, but accepts a few values to indicate default.

3) if you do `str_to_bool(val, default, ignore_error=True)`, then additionally every unparsable value is accepted and `default` is returned. This is the non-strict mode.

4) finally `str_to_bool(val, ignore_error=True)` is the same as `str_to_bool(val, False, ignore_error=True)`.


`default=None` wouldn't work, because `str_to_bool("default", None)` is valid and should give the default value `None`. The `DEFAULT_SENTINEL` is used to differenciate having the parameter unset. Under the assumption, that a caller would never use `str_to_bool(val, DEFAULT_SENTINEL)`. It's an internal tag.

Reply:
> default=None wouldn't work, because str_to_bool("default", None) is valid and should give the default value None.

Do you have an example of where that's currently valid?

Reply:
No. `_validate_bool()` doesn't. We should never allow `None` for a boolean value. If it's unset by the user then we use the defaults. In code we should never allow a "default" that's not `True`, `False`. Please fix this. It's a boolean; not a ternary.

Reply:
if the user writes
```
IndividualCalls=
IPv6_rpfilter=default
LockDown=-1
```
the their intent is pretty clear (to use the default). The current patch would handle that accordingly.

Of course, by dropping the `, default=None`, it will reject such values as invalid. Seems not very useful. This is known to be a boolean. If the user types `IndividualCalls=` then logging a warning to indicate that they did something wrong, is only annoying, not helpful. The behavior is almost the same except there won't be a warning about anything wrong.

Reply:
this is now on a separate commit `improvement(firewalld_conf): accept "", "-1", and "default" for boolean properties in "firewalld.conf"` , which can also be dropped...